From 255a750d28cd45df8923c3aba1e82c322757a50d Mon Sep 17 00:00:00 2001
From: Bruce Ashfield <bruce.ashfield@gmail.com>
Date: Tue, 8 Oct 2019 13:15:46 +0000
Subject: [PATCH] signal/ptrace: fix cgroup2/freezer long runtimes

As reported in the thread: https://lkml.org/lkml/2019/10/1/789, in
kernels with commit 76f969e8948d82 [cgroup: cgroup v2 freezer], we
were seeing much longer runtime in strace/ptrace tests (4 minutes
versus 4 seconds).

The issue only manifests if CONFIG_PREEMPT is enabled, which is in
all of the default configurations.

As sugggested in the thread, the movement of preempt_enable_no_resched()
until after the cgroup is frozen returns the behaviour to pre-5.2
runtimes.

Signed-off-by: Bruce Ashfield <bruce.ashfield@gmail.com>
---
 kernel/signal.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/signal.c b/kernel/signal.c
index 8aad311cbd59..07da8def4329 100644
--- a/kernel/signal.c
+++ b/kernel/signal.c
@@ -2155,8 +2155,8 @@ static void ptrace_stop(int exit_code, int why, int clear_code, kernel_siginfo_t
 		 */
 		preempt_disable();
 		read_unlock(&tasklist_lock);
-		preempt_enable_no_resched();
 		cgroup_enter_frozen();
+		preempt_enable_no_resched();
 		freezable_schedule();
 		cgroup_leave_frozen(true);
 	} else {
-- 
2.19.1

